docs/jetting.md

```diff
@@ -0,0 +1,709 @@
+```
+	XXX
+```
+There are two facets here: (i) data format, (ii) memory allocation.
+
+Regarding data format, if the hoon is expected to return a single atom (e.g. if the hoon looks like this
+```
+      ++  make-k            
+        ~/  %make-k
+        |=  [has=@uvI prv=@]     ::  <---- input parguments
+        ^-  @                    ::  <---- return value is a single value of type '@' (atom)
+		...
+```
+then your C code - at least when you're stubbing it out - can do something like
+```
+	 return(123);
+```
+Or, if you want to create an atom more formally, you can build it like this
+```
+     // this variable is on the stack and will disappear
+	 unsigned char nonce32[32];  
+
+	 // this allocates an indirect (> 31 bits) atom in the loom, 
+	 // does appropriate reference count, and returns the 32 bit handle to the atom
+     u3_noun nonce = u3i_words(8, (const c3_w*) nonce32);
+
+	 // this returns the 32 bit handle to the atom
+	 return(nonce);
+```
+If, on the other hand, your hoon looks like
+```
+      ++  ecdsa-raw-sign                                ::  generate signature
+        ~/  %ecdsa-raw-sign
+        |=  [has=@uvI prv=@]
+        ^-  [v=@ r=@ s=@]
+        ...
+```
+ending your C code with
+```
+	 return(123);
+```
+is wrong and will result in a dojo error because you are returning a single atom, instead of a list of three atoms.
+
+Instead do one of these:
+```
+     return(u3nc(a, b));		// for two atoms 
+     return(u3nt(a, b, c));		// for three atoms 
+     return(u3nq(a, b, c, d));	// for four atoms 
+```
+If you need to return a longer list, you can compose your own.  Look
+at the definitions of these three functions and you will see that they
+are just recursive calls to the cell constructor u3i_cell() e.g.
+```
+		 u3i_cell(a, u3i_cell(b, u3i_cell(c, d));
+```
+
+XXX more here about memory model, allocation, ownership
+
+
+
+## Compile the C code
+
+You've edited C code; now you need to compile it to build a new urbit executable!
+
+
+1. cd ~/tlon/urbit
+1. rm -rf build
+1. configure: choose either
+    1. meson configure -Dbuildtype=debug build
+    1. meson configure -Dbuildtype=release build
+	Note that debug builds run much much slower.  Booting a debug builkd of urbit may take several minutes.
+1. scripts/build
+
+## Compile the Hoon code into a pill
+
+You've compiled the Hoon code.  Now you need to compile it.
+
+Wait, compile?  Isn't Hoon that one types at the dojo prompt interpretted, or compiled on the fly?
+
+Yes.  But.
+
+When the urbit executable runs the first thing it does is load the
+complete arvo operating system.  That step is much faster if it can
+load a "jam"-ed "pill", where all of the hoon has already been parsed
+from text file into Hoon abstract syntax tree, and then compiled from
+the hoon into the Nock equivalent.
+
+Note that this means that if you edit hoon.hoon, zuse.hoon, arvo.hoon,
+vane/ames.hoon, etc. and then restart the urbit executable **YOU ARE
+NOT RUNNING YOUR NEW CODE**.
```

Comment:
I think this section and the following could be revised to be a little clearer. I'm not sure exactly what I'd changed, but here's some things to note about the pill (and `-A`):

An urbit has to boot into the Arvo kernel -- a nock core with a particular interface. It'd be possible to make some ad-hoc procedure to initialize Arvo, but it would be a pretty drastic layering violation and couple urbit to all sorts of internal implementation details of Arvo and Hoon. In contrast, the pill is basically a serialized, declarative set of steps to initialize Arvo.

Right now, the boot process is something of a hack, and the pill is basically a snapshot of a pristine Arvo kernel. In the future (see `+metal` or `+brass`), the pill will contain something like:

- a boot formula
- the hoon compiler as nock
- the hoon compiler as source
- arvo as source
- the vanes as source
- userspace as source (ie, an initial %clay sync)

For now, when you boot a ship that's not a galaxy, the initial %clay filesystem is synced in from the parent ship. When you boot a galaxy, you need to pass in a path to a copy of the Arvo repository so that the %clay filesystem can be initialized without recourse to another ship.

I'm not sure how much of that needs to be included (and there's potentially lots more), but I hope that helps to demystify these startup procedures.

Reply:
It's pushing 6pm on a Friday and I don't understand this perfectly myself, so I'm ADDING you version alongside my version.

Reply:
Fully understood. This is an incomplete and handwavy summary of a complex topic.

For future reference, `+brass` is the generator for the new pill (https://github.com/urbit/arvo/blob/master/gen/brass.hoon). Sections 3 and 4 of the whitepaper are highly relevant (http://media.urbit.org/whitepaper.pdf).